# Phase 3 Project

Student: Joshua Ruggles<br>
Instructor: Joe Comeaux/Morgan Jones<br>
Data Science Flex

<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Business-Problem" data-toc-modified-id="Business-Problem-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Business Problem</a></span><ul class="toc-item"><li><span><a href="#Ever-heard-of-'acabion'?" data-toc-modified-id="Ever-heard-of-'acabion'?-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Ever heard of 'acabion'?</a></span></li></ul></li><li><span><a href="#Brands-you've-never-heard-of-aside,-what-are-we-looking-for?" data-toc-modified-id="Brands-you've-never-heard-of-aside,-what-are-we-looking-for?-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Brands you've never heard of aside, what are we looking for?</a></span></li><li><span><a href="#What-makes-a-rating,-good?" data-toc-modified-id="What-makes-a-rating,-good?-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>What makes a rating, good?</a></span><ul class="toc-item"><li><span><a href="#Yamaha-MT-07-&amp;-Suzuki-SV650" data-toc-modified-id="Yamaha-MT-07-&amp;-Suzuki-SV650-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Yamaha MT-07 &amp; Suzuki SV650</a></span></li><li><span><a href="#Binary-values-for-'Rating'" data-toc-modified-id="Binary-values-for-'Rating'-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Binary values for 'Rating'</a></span></li></ul></li><li><span><a href="#Preprocessing-:-Expert-opinion" data-toc-modified-id="Preprocessing-:-Expert-opinion-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Preprocessing : Expert opinion</a></span><ul class="toc-item"><li><span><a href="#Dummy-categorical-columns" data-toc-modified-id="Dummy-categorical-columns-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Dummy categorical columns</a></span></li><li><span><a href="#train_test_split" data-toc-modified-id="train_test_split-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>train_test_split</a></span></li></ul></li><li><span><a href="#Baseline-Models" data-toc-modified-id="Baseline-Models-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Baseline Models</a></span><ul class="toc-item"><li><span><a href="#Baseline-Model-1.0:" data-toc-modified-id="Baseline-Model-1.0:-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Baseline Model 1.0:</a></span></li><li><span><a href="#KNN:-K-Nearest-Neighbors" data-toc-modified-id="KNN:-K-Nearest-Neighbors-5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>KNN: K-Nearest-Neighbors</a></span></li><li><span><a href="#BaggingClassifier()-1.0:" data-toc-modified-id="BaggingClassifier()-1.0:-5.3"><span class="toc-item-num">5.3&nbsp;&nbsp;</span>BaggingClassifier() 1.0:</a></span></li><li><span><a href="#BaggingClassifier()-2.0:" data-toc-modified-id="BaggingClassifier()-2.0:-5.4"><span class="toc-item-num">5.4&nbsp;&nbsp;</span>BaggingClassifier() 2.0:</a></span></li><li><span><a href="#Feature-Importance:-RandomSearchCV-3.0" data-toc-modified-id="Feature-Importance:-RandomSearchCV-3.0-5.5"><span class="toc-item-num">5.5&nbsp;&nbsp;</span>Feature Importance: RandomSearchCV 3.0</a></span></li></ul></li><li><span><a href="#The-Model" data-toc-modified-id="The-Model-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>The Model</a></span><ul class="toc-item"><li><span><a href="#Harley-Davidson-Street-500:-Testing-a-'bad'-motorcycle" data-toc-modified-id="Harley-Davidson-Street-500:-Testing-a-'bad'-motorcycle-6.1"><span class="toc-item-num">6.1&nbsp;&nbsp;</span>Harley-Davidson Street 500: Testing a 'bad' motorcycle</a></span><ul class="toc-item"><li><span><a href="#Preprocessing-our-model-for-the-Fi-Cruiser" data-toc-modified-id="Preprocessing-our-model-for-the-Fi-Cruiser-6.1.1"><span class="toc-item-num">6.1.1&nbsp;&nbsp;</span>Preprocessing our model for the Fi Cruiser</a></span></li></ul></li><li><span><a href="#Fi's-Cruiser,-but-let's-make-it-good" data-toc-modified-id="Fi's-Cruiser,-but-let's-make-it-good-6.2"><span class="toc-item-num">6.2&nbsp;&nbsp;</span>Fi's Cruiser, but let's make it good</a></span></li></ul></li></ul></div>

# Project Overview

For this project we will look at feature importance in the rating of a proposed new motorcycle. Specifically we will be using classification tools from sklearn. We will be utilizing the "all_bikez_curated.csv" file from Kaggle, curated by Emmanuel F. Werr

## Business Problem
Our data science team has been contracted by a fairly new up and coming motorcycle manufacturer, Fi Motors who is struggling to find their new formula for a 'fan favorite' motorcycle. 
<br><br>
Being a company of motorcycle riders, they have tasked our team to come up with a plan to replicate the successes of platforms like the Yamaha MT-07 and the Suzuki SV650. Fi Motor's ultimate desire, figure out what makes a motorcycle that people want to ride.


<img src="pictures/Andrew-SV650-1.webp"  />
<center>Tarmac to track: The SV650 affordably does it all<br>
Suzuki SV650 on track, courtesy of Bennetts Insurance
</center>

Further reading: 

Yamaha MT-07:<br> https://www.motorcyclenews.com/bike-reviews/yamaha/mt-07/2021/<br><br>https://www.motorcyclenews.com/bike-reviews/yamaha/mt-07/2014/<br><br>
https://www.youtube.com/watch?v=RUp2_7RhZEo

Suzuki SV650:<br> https://www.44teeth.com/suzuki-sv650-best-bike-ever/ <br><br>https://timeless2wheels.com/604079/suzuki-sv650/

In [2]:
# Import warnings 
import warnings
warnings.filterwarnings('ignore')

# Import pandas to read our data
import pandas as pd
# 'all_bikez_curated.csv', with a 'z'
df = pd.read_csv('all_bikez_curated.csv')
# Show our data 
df.head()

FileNotFoundError: [Errno 2] File b'all_bikez_curated.csv' does not exist: b'all_bikez_curated.csv'

In [3]:
df['Front suspension'].unique()

NameError: name 'df' is not defined

In [2]:
'''
In order to save us some effort down the road during the prediction phase of this notebook, let's 
create another set of vanilla data. 
'''
ds = pd.read_csv('all_bikez_curated.csv')

### Ever heard of 'acabion'? 

<img src="pictures/acabion_2.jpg" />
<center>Acabion GTBO450 (Not listed in dataset)</center>

Let's look at the unique values for 'Brand' as it would immediately appear that we probably aren't familiar with all of these motorcycles. Also, at ~700 hp, this thing seems more plane than bike.

In [3]:
# .unique() will give you a view of all of the potential outcomes in 'Brand'
df['Brand'].unique()

array(['acabion', 'access', 'ace', 'adiva', 'adler', 'adly', 'aeon',
       'aermacchi', 'agrati', 'ajp', 'ajs', 'alfer', 'alligator',
       'allstate', 'alphasports', 'alta', 'amazonas', 'american eagle',
       'american ironhorse', 'apc', 'aprilia', 'apsonic', 'arch',
       'arctic cat', 'ardie', 'ariel', 'arlen ness', 'arqin', 'askoll',
       'aspes', 'ather', 'atk', 'atlas honda', 'aurora',
       'avanturaa choppers', 'avinton', 'avon', 'azel', 'bajaj', 'balkan',
       'baltmotors', 'bamx', 'baotian', 'barossa', 'batavus', 'beeline',
       'benelli', 'bennche', 'beta', 'better', 'big bear choppers',
       'big dog', 'bimota', 'bintelli', 'black douglas', 'blackburne',
       'blata', 'bluroc', 'bmc choppers', 'bmw', 'boom trikes', 'borile',
       'boss hoss', 'bourget', 'bpg', 'brammo', 'bridgestone', 'britten',
       'brixton', 'brockhouse', 'brough superior', 'brudeli', 'bsa',
       'buccimoto', 'buell', 'bullit', 'bultaco', 'cagiva',
       'california scooter', 'can-

## Brands you've never heard of aside, what are we looking for? 

Though there have historically been a huge number of manufacturers with varying degrees of success, we're specifically looking for diamonds, regardless of country of origin, make, or model: if our goal is critical acclaim, let's then look at what the critics say, but let's first make sure that the critics actually said something. 

In [4]:
# Drop any null values in 'Rating'
df = df.dropna(subset=['Rating'])

In [5]:
# Preview first five rows and admire our work
df.head()

Brand                Model  Year                   Category  Rating  \
0   acabion      da vinci 650-vi  2011  Prototype / concept model     3.2   
1   acabion              gtbo 55  2007                      Sport     2.6   
2   acabion  gtbo 600 daytona-vi  2011  Prototype / concept model     3.5   
4   acabion              gtbo 70  2007  Prototype / concept model     3.1   
13   access      ams 8.57 lux lv  2016                        ATV     3.9   

    Displacement (ccm)  Power (hp)  Torque (Nm)  Engine cylinder  \
0                  NaN       804.0          NaN         Electric   
1               1300.0       541.0        420.0     In-line four   
2                  NaN       536.0          NaN         Electric   
4               1300.0       689.0        490.0     In-line four   
13               781.0        56.3          NaN  Single cylinder   

   Engine stroke  ... Dry weight (kg)  Wheelbase (mm) Seat height (mm)  \
0       Electric  ...           420.0             NaN              NaN   
1    four-stroke  ...           360.0             NaN              NaN   
2       Electric  ...           420.0             NaN              NaN   
4    four-stroke  ...           300.0             NaN              NaN   
13   four-stroke  ...           362.0             NaN              NaN   

    Front brakes  Rear brakes Front tire  Rear tire Front suspension  \
0    Single disc  Single disc        NaN        NaN              NaN   
1            NaN          NaN        NaN        NaN              NaN   
2    Single disc  Single disc        NaN        NaN              NaN   
4            NaN          NaN        NaN        NaN              NaN   
13   Single disc  Single disc   25/8-14   25/10-14               NaN   

    Rear suspension      Color options  
0               NaN                NaN  
1               NaN                NaN  
2               NaN                NaN  
4               NaN       Custom made.  
13              NaN  Black, white, red  

[5 rows x 28 columns]

<div class="alert alert-block alert-info">
There are a few things that we should look into prior to making a train_test_split; what do we believe will contribute to a higher rating for a motorcycle? Is there any data here that we can identify that may or may not give us those answers? Specifically we should see if we can make any sense out of all of the null values in this dataset. 
</div>

## What makes a rating, good? 

In [6]:
# .unique() will give you all of the potential scores found in the 'Rating' column 
df['Rating'].unique()

array([3.2, 2.6, 3.5, 3.1, 3.9, 2.8, 2.9, 1.9, 3.4, 2.2, 3.3, 3. , 2.7,
       2.5, 3.7, 2. , 3.6, 3.8, 2.1, 2.4, 4. , 4.1, 2.3, 1.8, 4.2, 1.7,
       4.4, 1.4, 4.3, 4.6, 1.6, 4.5, 1.5])

<div class="alert alert-block alert-info">
Since it would appear that our ratings work on a 5 point system, let's take a look at the motorcycles whose rating performance we would like to replicate: the Yamaha MT-07 and the Suzuki SV650.
</div>

### Yamaha MT-07 & Suzuki SV650

If like me, you are an avid motorcycle rider, you know, have known, or are someone on one of these bikes. 

<center><img src="pictures/mt07.jpg" alt="Drawing" style="width: 500px;"/> </center>

<center>MT-07, courtesy of Yamaha Powersports</center>
<!-- <img src="pictures/yamaha mt07.jfif" />
<img src="pictures/suzuki sv650.jfif" /> -->


In [7]:
# Find all included models that are an MT-07
df_yamaha = df.loc[df['Model'] == 'mt-07']
df_yamaha.head(1)

Brand  Model  Year    Category  Rating  Displacement (ccm)  \
35463  yamaha  mt-07  2014  Naked bike     3.5               689.0   

       Power (hp)  Torque (Nm) Engine cylinder Engine stroke  ...  \
35463        47.3         87.5            Twin   four-stroke  ...   

      Dry weight (kg)  Wheelbase (mm) Seat height (mm)  \
35463           175.0          1440.0            815.0   

                   Front brakes               Rear brakes    Front tire  \
35463  Double disc. Hydraulic.   Single disc. Hydraulic.   120/70-ZR17    

          Rear tire  Front suspension                   Rear suspension  \
35463  180/55-ZR17   Telescopic forks  Swingarm, (Link type suspension)   

       Color options  
35463   Black, white  

[1 rows x 28 columns]

In [8]:
# Find all models that have the name SV650
df_suzuki = df.loc[df['Model'] == 'sv650']
df_suzuki.head(1)

Brand  Model  Year    Category  Rating  Displacement (ccm)  \
30500  suzuki  sv650  2008  Naked bike     3.9               645.0   

       Power (hp)  Torque (Nm) Engine cylinder Engine stroke  ...  \
30500         NaN          NaN              V2   four-stroke  ...   

      Dry weight (kg)  Wheelbase (mm) Seat height (mm)  Front brakes  \
30500           168.0          1440.0            800.0   Double disc   

       Rear brakes    Front tire     Rear tire  \
30500  Single disc  120/60-ZR17   160/60-ZR17    

                                        Front suspension  \
30500  Telescopic, coil spring, oil damped, fully adj...   

                                  Rear suspension  Color options  
30500  Link-type, 7-way adjustable spring preload     Blue, Gray  

[1 rows x 28 columns]



Though this form factor has been out for nearly 20 years, the SV650 proves to still have relevance. 
<br><br>
<center><img src="pictures/Screenshot_20230123_110948.png"  style="width: 1000px;"/> </center>
<br>
<center>For the sake of posterity this was posted 1/21/2023</center>
<br>

In [9]:
# Bring in numpy
import numpy as np

In [10]:
# Average of ratings scores for the MT-07
yamaha_mean = np.mean(df_yamaha['Rating'].unique())
print('The average score of all year-model MT-07 in dataset:{}'.format(yamaha_mean))

The average score of all year-model MT-07 in dataset:3.35


In [11]:
# Average of ratings scores for the SV650
suzuki_mean = np.mean(df_suzuki['Rating'].unique())
print('The average score of all year-model SV650 in dataset:{}'.format(suzuki_mean))

The average score of all year-model SV650 in dataset:3.457142857142857


In [12]:
# Nothing fancy here, find the average of the two scores
rating_success = (yamaha_mean + suzuki_mean) / 2
print('Ultimately, this will be our ratings goal:{}'.format(rating_success))

Ultimately, this will be our ratings goal:3.4035714285714285


### Binary values for 'Rating'
Now that we have an average score for 'Rating', let's make a new column that will give us binary results, 1 being a rating above 3.4, 0 being a rating below. 

In [13]:
# Let's look for all of the unique values for the column 'Rating'
df['Rating'].unique()

array([3.2, 2.6, 3.5, 3.1, 3.9, 2.8, 2.9, 1.9, 3.4, 2.2, 3.3, 3. , 2.7,
       2.5, 3.7, 2. , 3.6, 3.8, 2.1, 2.4, 4. , 4.1, 2.3, 1.8, 4.2, 1.7,
       4.4, 1.4, 4.3, 4.6, 1.6, 4.5, 1.5])

In [14]:
# As detailed above, anything lower than 3.4 is outside of our target
rating_num = np.where(df["Rating"]>=(3.4), 1, 0)

In [15]:
# Our "Rating" column will now return a '1' for bikes that got a rating of 3.4 or above, 0 for scores
# below 3.4
rating_num

array([0, 0, 1, ..., 1, 1, 0])

In [16]:
# In order to attach this data to our current table we will need to make it into a dataframe
data = pd.DataFrame(rating_num, columns = ['Rating_'])
data.head()

Rating_
0        0
1        0
2        1
3        0
4        1

In [17]:
# Drop 'Rating' column 
df = df.drop(columns = 'Rating', axis =1)
df.head()

Brand                Model  Year                   Category  \
0   acabion      da vinci 650-vi  2011  Prototype / concept model   
1   acabion              gtbo 55  2007                      Sport   
2   acabion  gtbo 600 daytona-vi  2011  Prototype / concept model   
4   acabion              gtbo 70  2007  Prototype / concept model   
13   access      ams 8.57 lux lv  2016                        ATV   

    Displacement (ccm)  Power (hp)  Torque (Nm)  Engine cylinder  \
0                  NaN       804.0          NaN         Electric   
1               1300.0       541.0        420.0     In-line four   
2                  NaN       536.0          NaN         Electric   
4               1300.0       689.0        490.0     In-line four   
13               781.0        56.3          NaN  Single cylinder   

   Engine stroke    Gearbox  ...  Dry weight (kg) Wheelbase (mm)  \
0       Electric        NaN  ...            420.0            NaN   
1    four-stroke    6-speed  ...            360.0            NaN   
2       Electric        NaN  ...            420.0            NaN   
4    four-stroke    6-speed  ...            300.0            NaN   
13   four-stroke  Automatic  ...            362.0            NaN   

    Seat height (mm) Front brakes  Rear brakes Front tire  Rear tire  \
0                NaN  Single disc  Single disc        NaN        NaN   
1                NaN          NaN          NaN        NaN        NaN   
2                NaN  Single disc  Single disc        NaN        NaN   
4                NaN          NaN          NaN        NaN        NaN   
13               NaN  Single disc  Single disc   25/8-14   25/10-14    

    Front suspension  Rear suspension      Color options  
0                NaN              NaN                NaN  
1                NaN              NaN                NaN  
2                NaN              NaN                NaN  
4                NaN              NaN       Custom made.  
13               NaN              NaN  Black, white, red  

[5 rows x 27 columns]

In [18]:
# Put our new 'Rating_' column into our dataframe
df = pd.concat([df,data], axis =1)

In [19]:
# Make sure it worked
df.head()

Brand                Model    Year                   Category  \
0  acabion      da vinci 650-vi  2011.0  Prototype / concept model   
1  acabion              gtbo 55  2007.0                      Sport   
2  acabion  gtbo 600 daytona-vi  2011.0  Prototype / concept model   
3      NaN                  NaN     NaN                        NaN   
4  acabion              gtbo 70  2007.0  Prototype / concept model   

   Displacement (ccm)  Power (hp)  Torque (Nm) Engine cylinder Engine stroke  \
0                 NaN       804.0          NaN        Electric      Electric   
1              1300.0       541.0        420.0    In-line four   four-stroke   
2                 NaN       536.0          NaN        Electric      Electric   
3                 NaN         NaN          NaN             NaN           NaN   
4              1300.0       689.0        490.0    In-line four   four-stroke   

   Gearbox  ...  Wheelbase (mm) Seat height (mm)  Front brakes  Rear brakes  \
0      NaN  ...             NaN              NaN   Single disc  Single disc   
1  6-speed  ...             NaN              NaN           NaN          NaN   
2      NaN  ...             NaN              NaN   Single disc  Single disc   
3      NaN  ...             NaN              NaN           NaN          NaN   
4  6-speed  ...             NaN              NaN           NaN          NaN   

  Front tire Rear tire Front suspension  Rear suspension  Color options  \
0        NaN       NaN              NaN              NaN            NaN   
1        NaN       NaN              NaN              NaN            NaN   
2        NaN       NaN              NaN              NaN            NaN   
3        NaN       NaN              NaN              NaN            NaN   
4        NaN       NaN              NaN              NaN   Custom made.   

   Rating_  
0      0.0  
1      0.0  
2      1.0  
3      0.0  
4      1.0  

[5 rows x 28 columns]

In [20]:
# Preview data
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 30541 entries, 0 to 38461
Data columns (total 28 columns):
Brand                  21788 non-null object
Model                  21784 non-null object
Year                   21788 non-null float64
Category               21788 non-null object
Displacement (ccm)     21684 non-null float64
Power (hp)             14892 non-null float64
Torque (Nm)            10064 non-null float64
Engine cylinder        21778 non-null object
Engine stroke          21778 non-null object
Gearbox                19569 non-null object
Bore (mm)              17184 non-null float64
Stroke (mm)            17184 non-null object
Fuel capacity (lts)    18986 non-null float64
Fuel system            14546 non-null object
Fuel control           13444 non-null object
Cooling system         19740 non-null object
Transmission type      18547 non-null object
Dry weight (kg)        12705 non-null float64
Wheelbase (mm)         14347 non-null float64
Seat height (mm)       14460

<div class="alert alert-block alert-danger">
There are many missing values in our data currently, and also many different ways of dealing with those missing values. Datawig is a library that learns ML models using Deep Neural Networks to impute missing values in the data.
</div>

Our data is now ready for some preprocessing! 

## Preprocessing : Expert opinion
While there is still a lot of data (and we anticipate MANY dummied columns if we keep all of these selections), let's consider what things our manufacturer, Fi Motors does NOT care about when making a bike. <br><br>
____________________________________________________________________________________________________
<br>
After careful consideration from the team, it was decided that the following categories aren't incredibly important to Fi Motors as a manufacturer: <br><br>


- Category: <br><br>
There are several different types of vehicles in our data set, let's make sure that we aren't looking at anything that isn't a motorcycle.
<br><br>
- Fuel system: <br><br>
Most new bikes now are fuel injected (fuel *infected* if you're an old person) and do not rely on a carburettor. Since most current motorcycle riders would find trouble using a carburettor (save for museum pieces like The Yamaha TW200, Honda XR650L or Suzuki DRZ400), we intend on making a modern motorcycle with modern features. Fuel injection, it is. 
<br><br> 
- Gearbox: <br><br>
The only modern gearboxes are automatic (cvt transmissions like in the Honda Africa Twin or Rebel 1100), 5-speed (like on a Harley-Davidson) or 6-speed/6-speed with reverse (almost all modern sportbikes use 6-speed transmissions, reverse, again specifically for Honda, more specifically for the Goldwing). 6-speed transmissions are on both the Yamaha MT-07 and the SV650. 
<br><br>
- Fuel control: <br><br>
Fuel control most often comes down to SOHC (single overhead cam) or DOHC (dual overhead cam). Though there are differences, like Ducati's Desmodromic valve control system, most of the time the difference is writ elsewhere (like in horsepower and torque numbers). Ultimately this is a consideration in purchasing a bike (DOHC are often credited with producing more power out of the same displacement engine that a SOHC would) but not a huge consideration outside of racing. Even still, Fi is interested in seeing how fuel control might affect ratings. For the purposes of this model let's keep SOHC, DOHC, OHV, and the Desmodromic valve control system (because Ducati knows how to make a fun bike).   <br><br>

- Front brakes: <br><br>
Stopping and the ability to stop quickly is incredibly important for riders that find themselves on the track but likely the bike Fi intends on proposing will either have single or dual disc brakes; 7,451 of the bikes in our database utilize either single or dual disc brakes. Let's stick with these two selections.
<br><br>

- Rear brakes: <br><br>
The rear brakes are often seen as an afterthought for entry level bikes. The bike needs to be able to stop but it is incredibly unlikely that we will need to know how the premium parts rank versus a bike that has a stock braking set up for the rear wheel. Likely the rear brake will mirror the setting of the front brake. 
<br><br>

- Front tire: <br><br>
The height of the front tire is often a consideration with adventure motorcycles and dirt bikes because a taller tire can traverse larger obstacles. This is a sub-genre of motorcycle riding and will likely not have a huge impact on the enjoyability of a motorcycle. 
<br><br>
- Rear tire: <br><br>
Unlike the front tire, the rear tire is usually judged on its width; the wider the tire, usually the wider the tire the better traction the rear wheel will be able to put power down to the ground. Sportbikes usually have a wider contact patch on the sides of the tire, where drag bikes will sometimes even run a car tire. Ultimately the rear tire is a consideration to make when changing out rubber, but not necessarily something that will make or break your ride. 
<br><br>

- Front suspension: <br><br>
Suspension is incredibly important when it comes to a motorcycle and certainly having a branded suspension set up is an attractive selling point (think WP or Ohlins for example). However, there are too many variables for suspension set ups in our data to make heads or tails of if a bike with upgraded suspension will rank higher than a bike that does not. Furthermore, these are entry-level motorcycles we are looking to potentially manufacture. In order to keep costs down, usually the brake system and suspension systems are unbranded generic parts bought at bulk discount. 
<br><br>

- Rear suspension: <br><br>
For the same reasons given on front suspension, we will remove rear suspension from our data for now. 
<br><br>

- Color options: <br><br>
The paint job on a bike is usually a subjective metric and throughout different model years, the big three Japanese manufacturers will put out different color options (most launch models are available in two due to cost saving measures). 


In [21]:
# Drop categories that aren't actually a motorcycle
df = df[df['Category'].str.contains('Trial') == False]
df = df[df['Category'].str.contains('Speedway') == False]
df = df[df['Category'].str.contains('Minibike, cross') == False]
df = df[df['Category'].str.contains('ATV') == False]
df = df[df['Category'].str.contains('Scooter') == False]
df = df[df['Category'].str.contains('Unspecified category') == False]

In [22]:
# Check in on our 'Category' values
df['Category'].unique()

array(['Prototype / concept model', 'Sport', 'Cross / motocross',
       'Minibike, sport', 'Allround', 'Classic', 'Touring',
       'Enduro / offroad', 'Super motard', 'Custom / cruiser',
       'Naked bike', 'Sport touring'], dtype=object)

# Speedway racing
<img src="pictures/speedway motorcycle.jfif"  /><br><br>

With its skinny tires, low foot pegs, and racing DNA, we probably would not care how enjoyable a speedway bike is, regardless of how much fun it looks to race. 

In [23]:
# Drop unnecessary values from 'Gearbox'
df = df[df['Gearbox'].str.contains('1-speed') == False]
df = df[df['Gearbox'].str.contains('4-speed') == False]
df = df[df['Gearbox'].str.contains('3-speed') == False]
df = df[df['Gearbox'].str.contains('7-speed') == False]
df = df[df['Gearbox'].str.contains('4-speed with reverse') == False]
df = df[df['Gearbox'].str.contains('2-speed') == False]
df = df[df['Gearbox'].str.contains('10-speed') == False]
df = df[df['Gearbox'].str.contains('2-speed automatic') == False]
df = df[df['Gearbox'].str.contains('3-speed automatic') == False]
df = df[df['Gearbox'].str.contains('8-speed') == False]

In [24]:
# Drop columns listed above
df = df.drop(columns = 'Fuel system', axis = 1)
df = df.drop(columns = 'Rear brakes', axis =1)
df = df.drop(columns = 'Front tire', axis =1)
df = df.drop(columns = 'Rear tire', axis =1)
df = df.drop(columns = 'Front suspension', axis =1)
df = df.drop(columns = 'Rear suspension', axis =1)
df = df.drop(columns = 'Color options', axis =1)


In [25]:
# Remove commas from 'Stroke (mm)'
df['Stroke (mm)'] = df['Stroke (mm)'].str.replace(',','')

In [26]:
# Make 'Stroke (mm)' into a float
df['Stroke (mm)'] = df['Stroke (mm)'].astype('float64')

In [27]:
# View unique values for 'Fuel control'
df['Fuel control'].unique()

array([nan, 'Overhead Valves (OHV)', 'Overhead Cams (OHC)',
       'Single Overhead Cams (SOHC)',
       'Double Overhead Cams/Twin Cam (DOHC)', 'Rotary valves',
       'Membrane & variable discharge openings', 'Membrane',
       'Port control', 'Desmodromic valve control', 'Standing valves',
       'Membrane & power valve', 'Pushrods',
       'Rotary valves & variable discharge openings'], dtype=object)

In [28]:
# View instances of 'Fuel control'
df['Fuel control'].value_counts()

Double Overhead Cams/Twin Cam (DOHC)           4940
Single Overhead Cams (SOHC)                    1944
Overhead Cams (OHC)                            1305
Overhead Valves (OHV)                          1098
Desmodromic valve control                       497
Port control                                    259
Membrane                                        165
Rotary valves                                    65
Membrane & variable discharge openings           58
Membrane & power valve                           43
Pushrods                                         11
Rotary valves & variable discharge openings       3
Standing valves                                   2
Name: Fuel control, dtype: int64

In [29]:
# Fi asked us to look at DOHC, SOHC, OHC, and Desmodromic valve control, only. 
df = df[df['Fuel control'].str.contains('nan') ==False]
df = df[df['Fuel control'].str.contains('Port control') ==False]
df = df[df['Fuel control'].str.contains('Membrane') ==False]
df = df[df['Fuel control'].str.contains('Rotary valves') ==False]
df = df[df['Fuel control'].str.contains('Membrane & variable discharge openings') ==False]
df = df[df['Fuel control'].str.contains('Membrane & power valve') ==False]
df = df[df['Fuel control'].str.contains('Pushrods') ==False]
df = df[df['Fuel control'].str.contains('Rotary valves & variable discharge openings') ==False]
df = df[df['Fuel control'].str.contains('Standing valves') ==False]

In [30]:
# Preview the values for 'Front brakes'
df['Front brakes'].value_counts()

Single disc                                           2792
Double disc                                           1523
Dual disc                                             1353
Expanding brake                                        212
Double disc. ABS                                       183
                                                      ... 
Single disc. 2 piston                                    1
Double disc. Hydraulic disc. Optional ABS.               1
Double disc. Switchable ABS                              1
Single disc. 4-piston fixed                              1
Double disc. Twin floating disks, 4-piston caliper       1
Name: Front brakes, Length: 881, dtype: int64

In [31]:
df['Front brakes'].unique()

array(['Expanding brake', 'Single disc. 2 piston caliper',
       'Single disc. 2-piston calipers', 'Single disc',
       'Single disc. 2 piston calliper',
       'Single disc. double piston calliper',
       'Single disc. Brembo 2-piston calipers', 'Double disc',
       'Double disc. Floating discs',
       'Single disc. 4-piston polished calipers',
       'Double disc. 4-piston polished calipers', 'Double disc. ABS',
       'Double disc. ABS. Four-piston Brembo monoblock radial calipers.',
       'Double disc. ABS. Floating disk. Brembo M432 monoblock four-piston radial callipers.',
       'Double disc. Brembo four-piston radial callipers. Metal braided brake hose.',
       'Double disc. Bremo. ABS.',
       'Double disc. ABS.Stainless steel wave floating wave disc. 4-piston.',
       'Double disc. Bremo',
       'Double disc. Bremo stainless steel semi-floating disks.',
       'Double disc. ABS.Stainless steel floating disc. Radial calipers with four pistons. ',
       'Double disc.

Wow, there are a lot of different front brake configurations, 881 that exist in our dataset to be precise. We will only be looking at Single disc, Double disc, and Dual disc, so let's remove the others from our data.  

In [32]:
Single_brakes = df[df['Front brakes']== 'Single disc']
Dual_brakes = df[df['Front brakes'] == 'Dual disc']
Double_brakes = df[df['Front brakes'] == 'Double disc']

In [33]:
brakes = [Single_brakes, Dual_brakes, Double_brakes]
df = pd.concat(brakes)
df.head()

Brand                Model    Year          Category  Displacement (ccm)  \
358   ajp       pr4 125 enduro  2006.0  Enduro / offroad               124.0   
359   ajp       pr4 125 enduro  2007.0  Enduro / offroad               124.1   
362   ajp       pr4 125 enduro  2015.0  Enduro / offroad               124.0   
366   ajp  pr4 125 supermotard  2005.0      Super motard               125.0   
367   ajp  pr4 125 supermotard  2006.0      Super motard               124.0   

     Power (hp)  Torque (Nm)  Engine cylinder Engine stroke  Gearbox  ...  \
358        12.5          NaN  Single cylinder   four-stroke  5-speed  ...   
359        12.4          NaN  Single cylinder   four-stroke  5-speed  ...   
362        12.6          8.5  Single cylinder   four-stroke  5-speed  ...   
366        14.8          NaN  Single cylinder   four-stroke  5-speed  ...   
367        12.7          NaN  Single cylinder   four-stroke  5-speed  ...   

     Stroke (mm)  Fuel capacity (lts)                 Fuel control  \
358         49.5                  8.0  Single Overhead Cams (SOHC)   
359         49.5                  8.0  Single Overhead Cams (SOHC)   
362          NaN                  7.0          Overhead Cams (OHC)   
366          NaN                  9.0          Overhead Cams (OHC)   
367         49.5                  8.0  Single Overhead Cams (SOHC)   

    Cooling system Transmission type Dry weight (kg)  Wheelbase (mm)  \
358            Air             Chain           105.0          1410.0   
359            Air               NaN           105.0          1410.0   
362            Air             Chain           105.0          1380.0   
366            Air             Chain            99.0          1410.0   
367            Air               NaN           101.0          1410.0   

     Seat height (mm)  Front brakes Rating_  
358             910.0   Single disc     0.0  
359             910.0   Single disc     0.0  
362             920.0   Single disc     0.0  
366             850.0   Single disc     0.0  
367             905.0   Single disc     1.0  

[5 rows x 21 columns]

In [34]:
# Preview our data 
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5668 entries, 358 to 38273
Data columns (total 21 columns):
Brand                  5668 non-null object
Model                  5668 non-null object
Year                   5668 non-null float64
Category               5668 non-null object
Displacement (ccm)     5668 non-null float64
Power (hp)             3931 non-null float64
Torque (Nm)            1748 non-null float64
Engine cylinder        5668 non-null object
Engine stroke          5668 non-null object
Gearbox                5668 non-null object
Bore (mm)              5404 non-null float64
Stroke (mm)            4476 non-null float64
Fuel capacity (lts)    5468 non-null float64
Fuel control           5668 non-null object
Cooling system         5551 non-null object
Transmission type      5299 non-null object
Dry weight (kg)        2307 non-null float64
Wheelbase (mm)         3192 non-null float64
Seat height (mm)       3090 non-null float64
Front brakes           5668 non-null object


### Dummy categorical columns

Let's split our data into a categorical variable and dummy them into binary columns.

In [35]:
# Categorical/object columns can be found from df.info(), above
categoricals = ['Category','Engine cylinder', 'Engine stroke', 'Gearbox', 'Stroke (mm)', 
                'Cooling system', 'Transmission type', 'Fuel control', 'Front brakes'
                ]

# Our dataframe needs to be encoded. OHE can also be utilized, for this time we will use 
# pd.get_dummies()
df = pd.get_dummies(df, columns=categoricals)
# Preview our dummied columns
df.head()

Brand                Model    Year  Displacement (ccm)  Power (hp)  \
358   ajp       pr4 125 enduro  2006.0               124.0        12.5   
359   ajp       pr4 125 enduro  2007.0               124.1        12.4   
362   ajp       pr4 125 enduro  2015.0               124.0        12.6   
366   ajp  pr4 125 supermotard  2005.0               125.0        14.8   
367   ajp  pr4 125 supermotard  2006.0               124.0        12.7   

     Torque (Nm)  Bore (mm)  Fuel capacity (lts)  Dry weight (kg)  \
358          NaN       56.5                  8.0            105.0   
359          NaN       56.5                  8.0            105.0   
362          8.5        NaN                  7.0            105.0   
366          NaN        NaN                  9.0             99.0   
367          NaN       56.5                  8.0            101.0   

     Wheelbase (mm)  ...  Transmission type_Chain  \
358          1410.0  ...                        1   
359          1410.0  ...                        0   
362          1380.0  ...                        1   
366          1410.0  ...                        1   
367          1410.0  ...                        0   

     Transmission type_Shaft drive  Fuel control_Desmodromic valve control  \
358                              0                                       0   
359                              0                                       0   
362                              0                                       0   
366                              0                                       0   
367                              0                                       0   

     Fuel control_Double Overhead Cams/Twin Cam (DOHC)  \
358                                                  0   
359                                                  0   
362                                                  0   
366                                                  0   
367                                                  0   

     Fuel control_Overhead Cams (OHC)  Fuel control_Overhead Valves (OHV)  \
358                                 0                                   0   
359                                 0                                   0   
362                                 1                                   0   
366                                 1                                   0   
367                                 0                                   0   

     Fuel control_Single Overhead Cams (SOHC)  Front brakes_Double disc  \
358                                         1                         0   
359                                         1                         0   
362                                         0                         0   
366                                         0                         0   
367                                         1                         0   

     Front brakes_Dual disc  Front brakes_Single disc  
358                       0                         1  
359                       0                         1  
362                       0                         1  
366                       0                         1  
367                       0                         1  

[5 rows x 250 columns]

<div class="alert alert-block alert-success">
It looks like our columns were successfully dummied. Now we will need to split our data for modeling. 
</div>

In [36]:
df = df.dropna()

In [37]:
df['Rating_'].value_counts()

1.0    345
0.0    188
Name: Rating_, dtype: int64

### train_test_split


In [38]:
df = df.drop( columns = 'Brand', axis =1)
df = df.drop(columns = 'Model', axis =1)
df = df.drop(columns = 'Year', axis =1)


In [39]:
# target is 'y'
target = df['Rating_']
# X is the new name that we will use for our data. For testing purposes we may re-use 
#'df' if it produces better results.
X = df.drop(columns = 'Rating_', axis =1)

In [40]:
#import train_test_split
from sklearn.model_selection import train_test_split

# Create variables for modeling
X_train, X_test, y_train, y_test = train_test_split (X, target, random_state = 42)

In [48]:
# Because there is an error later in the code when running a confusion matrix, we will make a copy
# of all of the train_test_split data
X_trains = X_train.copy()
X_tests = X_test.copy()
y_trains = y_train.copy()
y_tests = y_test.copy()

## Baseline Models

### Baseline Model 1.0: 
First we will try a single DecisionTree

In [ ]:
#import DecisionTreeClassifier
from sklearn.tree import DecisionTreeClassifier 
dtc = DecisionTreeClassifier(random_state=42)  
# fit classifier onto training data
dtc.fit(X_train, y_train) 

In [ ]:
print('DecisionTree Model training score: {}'.format(dtc.score(X_train, y_train)))

In [ ]:
print('DecisionTree Model test score: {}'.format(dtc.score(X_test, y_test)))

<div class="alert alert-block alert-danger">
Our first model is built on a decision tree and the performance is not great; a training score that is nearly perfect paired with a test score of less than 60% would suggest that this first decision tree is overfitting to the data. This tree was created with no regard to hyperparameters, so let's call this our baseline model. In the notebook below we will see how we can improve it. 
</div>

### KNN: K-Nearest-Neighbors
KNN or K-Nearest-Neighbors is a distance based estimator; we will need to scale our data prior to modeling it in order to get best results. 


In [ ]:
# This next step can be done without instantiating a pipeline but the tool makes the process simpler
from sklearn.preprocessing import StandardScaler
from sklearn.neighbors import KNeighborsClassifier
from sklearn.pipeline import Pipeline
knn_pipeline = Pipeline([('ss', StandardScaler()),
                        ('knn', KNeighborsClassifier())])

In [ ]:
# Fit our knn_pipeline onto our training data, testing data and then see scores
knn_pipeline.fit(X_train,y_train)

print('KNN Training score:{}'.format(knn_pipeline.score(X_train,y_train)))
print('KNN Test score:{}'.format(knn_pipeline.score(X_test,y_test)))

Performance on our KNN model is not bad, but let's see if there is something that can give us a better test performance. 

### BaggingClassifier() 1.0: 

Bagging is a form of ensemble model training. If you think of planting a DecisionTree, this is like planting ANOTHER DecisionTree and using different parts of the data to improve our score.

In [ ]:
# import BaggingClassifier
from sklearn.ensemble import BaggingClassifier
b_tree =  BaggingClassifier(DecisionTreeClassifier(criterion='entropy', max_depth=6), 
                                 n_estimators=30, random_state = 42)
# fit model on training data
b_tree.fit(X_train, y_train)

In [ ]:
# Training data score
print('Bagged model training score: {}'.format(b_tree.score(X_train, y_train)))


In [ ]:
# Testing data score
print('Bagged model test score: {}'.format(b_tree.score(X_test, y_test)))


<div class="alert alert-block alert-success">
Our model has made improvements and does not appear to be overfitting. Since these are our best scores so far we should keep them in mind but for now let's move onto models 2.0.
</div>

### BaggingClassifier() 2.0: 

Let's lower n_estimators and max_depth

In [ ]:
BaggingClassifier().get_params().keys()

In [ ]:
b_tree1 =  BaggingClassifier(DecisionTreeClassifier(criterion='entropy', max_depth=5), 
                                 n_estimators=20, random_state = 42, bootstrap=True)
# fit model on training data
b_tree1.fit(X_train, y_train)

In [ ]:
print('Bagged model training score: {}'.format(b_tree1.score(X_train, y_train)))
print('Bagged model test score: {}'.format(b_tree1.score(X_test, y_test)))

<div class="alert alert-block alert-success">
Great! Our model has improved around ~1.5% on our testing data.
</div>

In [ ]:
X.info()

# DecisionTree Model 2.0: Cross-Validation

We will try using cross-validation to see if we can up the score of our base Decision Tree Model 1.0. Cross-validation assists with overfitting, and even though it did not appear that our model overfit the data, it is still worth a try.  

In [ ]:
# import cross_val_score from sklearn
from sklearn.model_selection import cross_val_score
# We will re-use 'dtc', our DecisionTreeClassifier() from earlier

dtc_cross_score = cross_val_score(dtc, X_train, y_train, cv = 5)
mean_dtc_cross_score = np.mean(dtc_cross_score)

print('Mean Cross-Validation Score:{}'.format(mean_dtc_cross_score))

<div class="alert alert-block alert-danger">
The performance on this model is worse than the first base DecisionTree that we put together. We will scrap this version for now and move forward with our Bagged Decision Tree. 
</div>

# RandomForestClassifier()
We can identify and interpret feature importance from our second ensemble model, RandomForestClassifier() 

In [ ]:
# import RandomForestClassifier
from sklearn.ensemble import RandomForestClassifier
# forest variable, random_state = 42, max_depth = 5
forest = RandomForestClassifier(random_state = 42, max_depth = 5)
# fit classifier on training data
forest.fit(X_train, y_train)

In [ ]:
print('RandomForest training score: {}'.format(forest.score(X_train, y_train)))

In [ ]:
print('RandomForest testing score: {}'.format(forest.score(X_test,y_test)))

<div class="alert alert-block alert-danger">
Though this would not appear to be a successful model, let's begin hyperparameter tuning to see if we can get a score of above, say 80% on both the training and testing data
</div>

# Hyperparameter tuning on RandomForestClassifier()

In [ ]:
RandomForestClassifier().get_params().keys()

In [ ]:
# Set a new variable for our second version of the RandomForestClassifier()
# As a reminder: 
# forest = RandomForestClassifier(random_state =42, max_depth = 5)
forest1 = RandomForestClassifier(random_state =42, max_depth =5, n_estimators=10)
    

In [ ]:
forest1.fit(X_train,y_train)
print('RandomForestClassifier training score n_estimators = 10: {}'.format(forest1.score(X_train, y_train)))
print('RandomForestClassifier test score n_estimators =10: {}'.format(forest1.score(X_test, y_test)))

In [ ]:
forest2 = RandomForestClassifier(criterion = 'gini',random_state = 42, max_depth =5, n_estimators =50)
forest2.fit(X_train,y_train)
print('RandomForestClassifier training score n_estimators = 50: {}'.format(forest2.score(X_train, y_train)))
print('RandomForestClassifier test score n_estimators =50: {}'.format(forest2.score(X_test, y_test)))

<div class="alert alert-block alert-danger">
Not exactly what we had in mind. We have created an ensemble model but are having a hard time determining parameters that will increase score. For this, let's use Random Search. 
</div>

# Random Search Training

In [ ]:
# random_grid works the same way that a param_grid would: we will see what parameters are selected as 
# the 'best' parameters and build our next model based on these results. 
random_grid = {'bootstrap': [True, False],
               'max_depth': [10,30,50,70,90,100, None],
               'max_features': ['auto', 'sqrt'], 
               'n_estimators': [200, 400, 600]    
}

In [ ]:
# This takes almost 20 minutes to run, Weird Al Yankovic's 'Sauerkraut' x 2 would be unbearable but...
from sklearn.model_selection import RandomizedSearchCV
rf = RandomForestClassifier()
rf_cv1 = RandomizedSearchCV(estimator =rf,
                          param_distributions = random_grid, 
                           n_iter = 100,
                          cv = 3
                          )
rf_cv1.fit(X_train, y_train)
rf_cv1.score(X_train, y_train)
# Expected output: 0.9573934837092731

In [ ]:
rf_cv1.score(X_test, y_test)
# Expected output: 0.6716417910447762

In [ ]:
# The best parameters in this first tuned model
rf_cv1.best_params_
# Expected output: {'n_estimators': 400,
#  'max_features': 'sqrt',
#  'max_depth': 30,
#  'bootstrap': True}

<div class="alert alert-block alert-danger">
Thinking critically on these parameters then, 600 estimators was the most in our set, let's shoot for more in the following model, max_features: 'auto', max_depth: '(something in the single digits)', and 'bootstrap': False
</div>

# Random Search Training 2.0
Again, taking into account the parameters that were selected as the 'best', let's expand on that idea by making our next set of parameters exploratory options referentially to our previous set of 'best_params'. 

In [ ]:
# Set new parameters for new random_grid1
random_grid1 = {'bootstrap': [True],
               'max_depth': [1,5,7,10],
               'max_features': ['auto'], 
               'n_estimators': [600,800,1000,1200]    
}

In [ ]:
rf_cv = RandomizedSearchCV(estimator =rf,
                          param_distributions = random_grid1, 
                           n_iter = 100,
                          cv = 3
                          )
rf_cv.fit(X_train, y_train)
rf_cv.score(X_train, y_train)
# Expected output: 0.7014634146341463

In [ ]:
rf_cv.score(X_test, y_test)
# Expected output: 0.6549707602339181

In [ ]:
rf_cv.best_params_
# Expected output: {'n_estimators': 600,
#  'max_features': 'auto',
#  'max_depth': 10,
#  'bootstrap': False}

<div class="alert alert-block alert-danger">
Well there's a slight improvement, and we have gotten closer to the best score we've gotten so far from our BaggedClassifier(DecisionTreeClassifier()) but let's keep tuning. 
</div>

# Random Search Training 3.0

In [ ]:
random_grid2 = {'bootstrap': [False],
               'max_depth': [10,11,12,13,14,15,16,17,18,19,20],
               'max_features': ['auto'], 
               'n_estimators': [700, 750, 800, 850, 900]    
}

In [ ]:
# 'Stairway to Heaven' like thrice
rf_cv = RandomizedSearchCV(estimator =rf,
                          param_distributions = random_grid2, 
                           n_iter = 100,
                          cv = 5 # Changed from 3 to 5, probably accounts for the time
                          )
rf_cv.fit(X_train, y_train)
rf_cv.score(X_train, y_train)
# Expected output: 0.7346341463414634

In [ ]:
rf_cv.score(X_test, y_test)
# Expected output: 0.652046783625731

In [ ]:
rf_cv.best_params_
#{'n_estimators': 900,
#  'max_features': 'auto',
#  'max_depth': 12,
#  'bootstrap': False}

# AdaboostClassifier
While our model is performing at a rate that could be considered passable, let's try using an AdaboostClassifier to see if we can improve a model. 

In [ ]:
# import AdaBoostClassifier from sklearn
from sklearn.ensemble import AdaBoostClassifier
AdaBoostClassifier().get_params().keys()

In [ ]:
# A pretty vanilla classifier to start
ada = AdaBoostClassifier(random_state = 42)

In [ ]:
ada.fit(X_train, y_train)
ada.score(X_train, y_train), ada.score(X_test, y_test)

In [ ]:
# ada1 is our first attempt at improving the vanilla AdaBoostClassifier()
ada1 = AdaBoostClassifier(random_state = 42, learning_rate = 1.0, n_estimators = 300 )
ada1.fit(X_train, y_train)
ada1.score(X_train, y_train), ada1.score(X_test, y_test)
# Expected output: (0.8907317073170732, 0.6432748538011696)

In [ ]:
# Trying out a learning_rate = 0.2, n_estimators = 300
ada2 = AdaBoostClassifier(random_state = 42, learning_rate = 0.1, n_estimators = 100 )
ada2.fit(X_train, y_train)
ada2.score(X_train, y_train), ada2.score(X_test, y_test)
# Expected output: (0.8126829268292682, 0.6491228070175439)

In [ ]:
# Using the proposed learning_rate = 0.000975 and n_estimators = 500
ada3 = AdaBoostClassifier(random_state = 42, learning_rate = 0.1, n_estimators = 200 )
ada3.fit(X_train, y_train)
ada3.score(X_train, y_train), ada3.score(X_test, y_test)
# Expected output: (0.6068292682926829, 0.6461988304093568)

In [ ]:
# n_estimators = 1500 to see if there is improvement by increasing n_estimators 
ada4 = AdaBoostClassifier(random_state = 42, learning_rate = 0.1, n_estimators = 150 )
ada4.fit(X_train, y_train)
ada4.score(X_train, y_train), ada4.score(X_test, y_test)
# Expected output: (0.7243107769423559, 0.6492537313432836)

### Feature Importance: RandomSearchCV 3.0

RandomSearchCV 3.0 was our best performant model: <br><br>

Training : 0.889<br>
Test: 0.694<br>

In [ ]:
rf_cv.get_params().keys()

In [ ]:
# Because RandomSearchCV does not have a feature importances function, let's try this:
rf_cv.best_estimator_.feature_importances_

In [ ]:
feat_importances = pd.Series(rf_cv.best_estimator_.feature_importances_, index=X.columns)
feat_importances.nlargest(10).plot(kind='barh', color = 'orange');



In [ ]:
xx

## The Model

So, we have done many things so far, but we will have to duplicate some of our efforts in order to get a prediction on whether or not a bike with certain features will be a ratings success or flop. Let's begin with the original data. 

In [ ]:
# ds will be our the dataset we fit our predictive model onto. 
ds = pd.read_csv('all_bikez_curated.csv')
ds.head()

In [ ]:
ds.info()

In [ ]:
# The only categorical column not included is 'Rating'
categorical = ['Brand', 'Model', 'Year', 'Category', 'Rating', 'Engine cylinder', 'Engine stroke', 
                'Gearbox', 'Stroke (mm)', 'Fuel control', 'Cooling system', 
               'Transmission type', 'Front brakes'
               ]

continuous = ['Displacement (ccm)', 'Power (hp)', 'Torque (Nm)', 'Bore (mm)', 'Fuel capacity (lts)',
             'Dry weight (kg)', 'Wheelbase (mm)', 'Seat height (mm)']

In [ ]:
# We will need to fill all of these below columns
used_cols = [*categorical, *continuous]
used_cols

### Harley-Davidson Street 500: Testing a 'bad' motorcycle

FI is our manufacturer's brand so let's use them for 'Brand', for now let's use motorcycle type as the 'Model'

In [ ]:
pd.set_option('max_columns', None)
df_500 = ds.loc[ds['Model'] == 'street 500']
df_500.head()

In [ ]:
new_row_cruiser = pd.DataFrame(columns = used_cols)

In [ ]:
# To test our model performance, let's pick a cruiser motorcycle that is sort of like the infamous
# Harley-Davidson Street 500, a bike in infamy because of the amount used for MSF courses 
new_row_cruiser = new_row_cruiser.append({
    'Brand': 'Harley-Davidson', 
    'Model': 'street 500',
    'Year': '2014',
    'Category': 'Custom / cruiser', 
    'Engine cylinder': 'V2', 
    'Engine stroke': ' four-stroke', 
    'Gearbox': '6-speed', 
    'Stroke (mm)': '66.0', 
    'Fuel capacity (lts)': 13.25, 
    'Fuel control': 'Single Overhead Cams (SOHC)', 
    'Cooling system': 'Liquid', 
    'Transmission type': 'Belt', 
    'Dry weight (kg)': 206.4, 
    'Wheelbase (mm)': 1534.0, 
    'Seat height (mm)': 653.0, 
    'Displacement (ccm)': 500,
    'Power (hp)': 50, 
    'Torque (Nm)': 40,
    'Bore (mm)': 69.1,
    'Front brakes': 'Single disc. Twin piston, floating caliper', 

    
},ignore_index = True)

In [ ]:
# To test the efficacy of our predictive model let's also include the Yamaha MT-07
new_row_cruiser = new_row_cruiser.append({
    'Brand': 'Yamaha', 
    'Model': 'MT-07',
    'Year': '2020',
    'Category': 'Naked bike', 
    'Engine cylinder': 'Twin', 
    'Engine stroke': ' four-stroke', 
    'Gearbox': '6-speed', 
    'Stroke (mm)': '68.6', 
    'Fuel capacity (lts)': 14.0, 
    'Fuel control': 'Double Overhead Cams/Twin Cam (DOHC)', 
    'Cooling system': 'Liquid', 
    'Transmission type': 'Chain', 
    'Dry weight (kg)': 175, 
    'Wheelbase (mm)': 1440.0, 
    'Seat height (mm)': 815.0, 
    'Displacement (ccm)': 689.0,
    'Power (hp)': 74.0, 
    'Torque (Nm)': 68.0,
    'Bore (mm)': 80.0,
    'Front brakes': 'Double disc', 

    
},ignore_index = True)

In [ ]:
# If we are including the MT-07, let's also include the SV650
new_row_cruiser = new_row_cruiser.append({
    'Brand': 'Suzuki', 
    'Model': 'SV650',
    'Year': '2022',
    'Category': 'Naked bike', 
    'Engine cylinder': 'V2', 
    'Engine stroke': ' four-stroke', 
    'Gearbox': '6-speed', 
    'Stroke (mm)': '62.6', 
    'Fuel capacity (lts)': 17.0, 
    'Fuel control': 'Double Overhead Cams/Twin Cam (DOHC)', 
    'Cooling system': 'Liquid', 
    'Transmission type': 'Chain', 
    'Dry weight (kg)': 168, 
    'Wheelbase (mm)': 1440.0, 
    'Seat height (mm)': 800.0, 
    'Displacement (ccm)': 645.0,
    'Power (hp)': 75.0, 
    'Torque (Nm)': 64.0,
    'Bore (mm)': 81.0,
    'Front brakes': 'Double disc', 

    
},ignore_index = True)

In [ ]:
# Let's also include some new bikes based on feature importance, how about a cruiser?
df_kawasaki = ds.loc[ds['Model'] == 'vulcan s']
df_kawasaki.head(5)


In [ ]:
new_row_cruiser = new_row_cruiser.append({
    'Brand': 'Fi', 
    'Model': 'Hephaestus', # In Roman mythology, Vulcan is the god of fire and metalwork. 
    'Year': '2022',
    'Category': 'Custom / cruiser', 
    'Engine cylinder': 'Twin', 
    'Engine stroke': ' four-stroke', 
    'Gearbox': '6-speed', 
    'Stroke (mm)': '60.0', 
    'Fuel capacity (lts)': 14.0, 
    'Fuel control': 'Double Overhead Cams/Twin Cam (DOHC)', 
    'Cooling system': 'Liquid', 
    'Transmission type': 'Chain', 
    'Dry weight (kg)': 168, # There is no weight in our data for this bike, but let's improve by using SV650 weight
    'Wheelbase (mm)': 1575.0, 
    'Seat height (mm)': 705.0, 
    'Displacement (ccm)': 649.0,
    'Power (hp)': 60.4, 
    'Torque (Nm)': 63.0,
    'Bore (mm)': 83.0,
    'Front brakes': 'Double disc', 

    
},ignore_index = True)

In [ ]:
# What about an adventure bike that is worth improving? Let's see about the BMW GS1200
df_bmw = ds.loc[ds['Model'] == 'r1200gs']
df_bmw.head(5)

In [ ]:
# There is only one row for BMW R1200GS and it is rated at a 3.8
new_row_cruiser = new_row_cruiser.append({
    'Brand': 'Fi', 
    'Model': 'R1300GS', # Though there is an R1250GS in production currently, let's put an even larger engine
    'Year': '2022',
    'Category': 'Enduro / offroad', 
    'Engine cylinder': 'Twin cylinder boxer', # The legendary and iconic boxer engine
    'Engine stroke': ' four-stroke', 
    'Gearbox': '6-speed', 
    'Stroke (mm)': '101.0', 
    'Fuel capacity (lts)': 20.0, 
    'Fuel control': 'Single Overhead Cams (SOHC)', 
    'Cooling system': 'Liquid', 
    'Transmission type': 'Chain', 
    'Dry weight (kg)': 175, # The original R1200GS weighs 199.0 kg
    'Wheelbase (mm)': 1519.0, 
    'Seat height (mm)': 840.0, 
    'Displacement (ccm)': 1299.0,
    'Power (hp)': 120.0, # Original is 102.0 
    'Torque (Nm)': 130.0, # Original is 115.0
    'Bore (mm)': 101.0,
    'Front brakes': 'Double disc', 

    
},ignore_index = True)

In [ ]:
# With its flat track roots, the FTR1200 is considered one of the most beautiful bikes in production
# 
df_indian = ds.loc[ds['Model'] == 'ftr 1200']
df_indian.head(5)

In [ ]:
# There is only one row for BMW R1200GS and it is rated at a 3.8
new_row_cruiser = new_row_cruiser.append({
    'Brand': 'Fi', 
    'Model': 'Fi1200', # Though there is an R1250GS in production currently, let's put an even larger engine
    'Year': '2019',
    'Category': 'Custom / cruiser', 
    'Engine cylinder': 'V2',
    'Engine stroke': ' four-stroke', 
    'Gearbox': '6-speed', 
    'Stroke (mm)': '73.6', 
    'Fuel capacity (lts)': 12.9, 
    'Fuel control': 'Double Overhead Cams/Twin Cam (DOHC)', 
    'Cooling system': 'Liquid', 
    'Transmission type': 'Chain', # The original bike is belt driven
    'Dry weight (kg)': 220.0, # The original FTR1200 weighs 230.0 kg
    'Wheelbase (mm)': 1524.0, 
    'Seat height (mm)': 840.0, 
    'Displacement (ccm)': 1299.0, # The original displacement is 1203.0
    'Power (hp)': 123.0, 
    'Torque (Nm)': 130.0, # Original is 117.0
    'Bore (mm)': 102.0,
    'Front brakes': 'Double disc', 

    
},ignore_index = True)

In [ ]:
new_row_cruiser

#### Preprocessing our model for the Fi Cruiser

In [ ]:
xx

In [ ]:
nrc = new_row_cruiser
y_nrc = nrc['Rating']

In [ ]:
y_nrc = np.where(y_nrc >= (3.4), 1, 0)

In [ ]:
y_nrc

In [ ]:
nrc = nrc.drop(columns = 'Rating', axis = 1)
nrc.head(6)

In [ ]:
nrc = pd.get_dummies(nrc, columns = categoricals)
nrc.head(6)

In [ ]:
nrc = nrc.drop(columns = 'Brand', axis =1)
nrc = nrc.drop(columns = 'Year', axis =1)
nrc = nrc.drop(columns = 'Model', axis =1)

In [ ]:
# Because we used 5 splits (cv=5), you will need to have at LEAST 5 bikes to sample from. You won't
# have to use them all, but you'll have to provide at least 5. 
rf_cv.fit(nrc, y_nrc)

In [ ]:
rf_cv.predict(nrc)

### Fi's Cruiser, but let's make it good

In [ ]:
fi_cruiser = pd.DataFrame(columns = used_cols)

fi_cruiser = new_row_cruiser.append({
    'Brand': 'FI', 
    'Model': 'Cruiser',
    'Year': '2023',
    'Category': 'Custom / cruiser', 
    'Engine cylinder': 'V2', 
    'Engine stroke': ' four-stroke', 
    'Gearbox': '6-speed', 
    'Stroke (mm)': '66.0', 
    'Fuel capacity (lts)': 13.25, 
    'Fuel control': 'Single Overhead Cams (SOHC)', 
    'Cooling system': 'Liquid', 
    'Transmission type': 'Chain', 
    'Dry weight (kg)': 206.4, 
    'Wheelbase (mm)': 1534.0, 
    'Seat height (mm)': 653.0, 
    'Displacement (ccm)': 500,
    'Power (hp)': 50, 
    'Torque (Nm)': 40,
    'Bore (mm)': 69.1,
    'Front brakes': 'Single disc. Twin piston, floating caliper',
    'Front suspension': 'Telescopic fork'
    
},ignore_index = True)